samza-core/src/test/java/org/apache/samza/processor/TestStreamProcessor.java

```diff
@@ -165,10 +175,98 @@ public void onFailure(Throwable t) {
+   * through the StreamProcessor
+   *
+   * Assertions:
+   * - JobCoordinator has been stopped from the JobCoordinatorListener callback
+   * - StreamProcessorLifecycleListener#onFailure(Throwable) has been invoked
+   */
+  @Test
+  public void testContainerFailureCorrectlyStopsProcessor() {
+    JobCoordinator mockJobCoordinator = mock(JobCoordinator.class);
+    Throwable expectedThrowable =  new SamzaException("Failure in Container!");
+    AtomicReference<Throwable> actualThrowable = new AtomicReference<>();
+    final CountDownLatch runLoopStartedLatch = new CountDownLatch(1);
+    RunLoop failingRunLoop = mock(RunLoop.class);
+    doAnswer(invocation ->
+      {
+        try {
+          Thread.sleep(10);
+          runLoopStartedLatch.countDown();
+          throw expectedThrowable;
+        } catch (InterruptedException ie) {
+          ie.printStackTrace();
+        }
+        return null;
+      }).when(failingRunLoop).run();
+
+    SamzaContainer mockContainer = StreamProcessorTestUtils.getDummyContainer(failingRunLoop, mock(StreamTask.class));
+
+    final CountDownLatch processorListenerFailed = new CountDownLatch(1);
+
+    TestableStreamProcessor processor = new TestableStreamProcessor(
+        new MapConfig(),
+        new HashMap<>(),
+        mock(StreamTaskFactory.class),
+        new StreamProcessorLifecycleListener() {
+          @Override
+          public void onStart() { }
+
+          @Override
+          public void onShutdown() { }
+
+          @Override
+          public void onFailure(Throwable t) {
+            actualThrowable.getAndSet(t);
+            processorListenerFailed.countDown();
+          }
+        },
+        mockJobCoordinator,
+        mockContainer);
+
+    final CountDownLatch coordinatorStop = new CountDownLatch(1);
+    doAnswer(invocation ->
+      {
+        coordinatorStop.countDown();
+        return null;
+      }).when(mockJobCoordinator).stop();
+
+    doAnswer(invocation ->
+      {
+        new Thread(() ->
+          {
+            try {
+              processor.jobCoordinatorListener.onNewJobModel("1", getMockJobModel());
+              coordinatorStop.await();
+              processor.jobCoordinatorListener.onCoordinatorStop();
+            } catch (InterruptedException e) {
+              e.printStackTrace();
+            }
+          }).start();
+        return null;
+      }).when(mockJobCoordinator).start();
+
+    try {
+      processor.start();
+
+      // This block is required for the mockRunloop is actually start.
+      // Otherwise, processor.stop gets triggered before mockRunloop begins to block
+      runLoopStartedLatch.await();
+      Assert.assertTrue(
+          "Container failed and processor listener failed was not invoked within timeout!",
+          processorListenerFailed.await(30, TimeUnit.SECONDS));
+      Assert.assertEquals(expectedThrowable, actualThrowable.get());
+
+    } catch (InterruptedException e) {
```

Comment:
We should not be expecting the test to be interrupted here, so it would be reasonable to throw here instead of swallowing the exception.